packages/clay-drop-down/src/DropDownWithDrilldown.tsx

```diff
@@ -0,0 +1,212 @@
+/**
+ * SPDX-FileCopyrightText: © 2020 Liferay, Inc. <https://liferay.com>
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import ClayButton from '@clayui/button';
+import React from 'react';
+
+import ClayDropDown from './DropDown';
+import {ClayDropDownContext, findNested} from './DropDownWithItems';
+import Header from './Header';
+import Menu from './Menu';
+
+type TType = 'checkbox' | 'group' | 'item' | 'radio' | 'radiogroup' | 'divider';
+
+interface IItem {
+	active?: boolean;
+	checked?: boolean;
+	disabled?: boolean;
+	href?: string;
+	icon?: string;
+	childItems?: Array<IItem>;
+	items?: Array<IItem>;
+	label?: string;
+	name?: string;
+	onChange?: Function;
+	onClick?: (e: React.MouseEvent<HTMLElement, MouseEvent>) => void;
+	symbolLeft?: string;
+	symbolRight?: string;
+	type?: TType;
+	value?: string;
+}
+
+interface IDropDownContentProps {
+	/**
+	 * The path to the SVG spritemap file containing the icons.
+	 */
+	spritemap?: string;
+
+	/**
+	 * List of items to display in drop down menu
+	 */
+	items: Array<IItem>;
+}
+
+export interface IProps extends IDropDownContentProps {
+	/**
+	 * Default position of menu element. Values come from `./Menu`.
+	 */
+	alignmentPosition?: React.ComponentProps<typeof Menu>['alignmentPosition'];
+
+	/**
+	 * Informational text that appears at the end or above the `footerContent` prop.
+	 */
+	caption?: string;
+
+	className?: string;
+
+	/**
+	 * HTML element tag that the container should render.
+	 */
+	containerElement?: React.ComponentProps<
+		typeof ClayDropDown
+	>['containerElement'];
+
+	/**
+	 * A label for the button in the Footer, use this instead of footerContent if you want a single Button
+	 */
+	footerBtnLabel?: string;
+
+	/**
+	 * Add an action button or any other element you want to be fixed position to the
+	 * footer from the DropDown.
+	 */
+	footerContent?: React.ReactElement;
+
+	/**
+	 * Element that is used as the trigger which will activate the dropdown on click.
+	 */
+	trigger: React.ReactElement;
+
+	/**
+	 * Text to be displayed in the Dropdown's header.
+	 */
+	headerText?: string;
```

Comment:
I don't think I would expect users to have to manually handle providing the headerText and the undo callback. Ideally we would handle the back/forth logic internally for the users and all they need to add is the "children" attribute to the item they want to have drilldown abilities.

So if we have an item like this
```js
[{
    name: 'Documents and Media',
    children: [
        {
            name: 'Pictures',
            href: '#'
        }
    ]
}]
```

The item should automatically render a simple dropdown list with one item with an arrow and no header. If the user clicks the "Documents and Media" item, it should then render its children list of items, and the header would read "Documents and Media" and a back arrow. When the back arrow is clicked, it should go back to the original list.

Ideally we can handle all of that logic internally so that the user doesn't have to wire anything up themselves other than providing the items.

You may even try just updating `DropDownWithItems` first to see if you can add the logic needed there and not have to start entirely from scratch. This way you can focus just on the back/forth logic and interactions, rather than the menu as a whole.